src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -266,6 +264,96 @@ private void HandleChannelException(IConnection connection, Exception ex)
             }
         }
 
+        private bool IsNotMaster(ServerErrorCode code, string message)
+        {
+            switch (code)
+            {
+                case ServerErrorCode.NotMaster: // 10107
+                case ServerErrorCode.NotMasterNoSlaveOk: // 13435
+                    return true;
+            }
+
+            if (message != null)
+            {
+                if (message.IndexOf("not master", StringComparison.OrdinalIgnoreCase) != -1 &&
+                    message.IndexOf("not master or secondary", StringComparison.OrdinalIgnoreCase) == -1)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private bool IsNotMasterOrRecovering(ServerErrorCode code, string message)
+        {
+            return IsNotMaster(code, message) || IsRecovering(code, message);
+        }
+
+        private bool IsRecovering(ServerErrorCode code, string message)
+        {
+            switch (code)
+            {
+                case ServerErrorCode.InterruptedAtShutdown: // 11600
+                case ServerErrorCode.InterruptedDueToReplStateChange: // 11602
+                case ServerErrorCode.NotMasterOrSecondary: // 13436
+                case ServerErrorCode.PrimarySteppedDown: // 189
+                case ServerErrorCode.ShutdownInProgress: // 91
+                    return true;
+            }
+
+            if (message != null)
+            {
+                if (message.IndexOf("not master or secondary", StringComparison.OrdinalIgnoreCase) != -1 ||
+                    message.IndexOf("node is recovering", StringComparison.OrdinalIgnoreCase) != -1)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private bool ShouldInvalidateServer(Exception exception)
+        {
+            if (__invalidatingExceptions.Contains(exception.GetType()))
+            {
+                return true;
+            }
+
+            var commandException = exception as MongoCommandException;
+            if (commandException != null)
+            {
+                var code = (ServerErrorCode)commandException.Code;
+                var message = commandException.ErrorMessage;
+
+                if (IsNotMasterOrRecovering(code, message))
+                {
+                    return true;
+                }
+
+                if (commandException.GetType() == typeof(MongoWriteConcernException))
```

Comment:
If the top level code and message didn't merit returning true AND IF this is a MongoWriteConcernException then let's look at the writeConcernError code and message.